tools/wds2idx.py

```diff
@@ -0,0 +1,207 @@
+        Path to the index file, that will be created/overwritten.
+    """
+
+    tar_block_size = 512
+    index_file_version = "v1.0"
+
+    def __init__(self, uri, idx_path):
+        self.uri = uri
+        self.idx_path = idx_path
+        self.fidx = open(self.idx_path, "w")
+
+    def open(self):
+        """Opens the archive and index files and sets their read heads to 0."""
+        if self.fidx.closed:
+            self.fidx = open(self.idx_path, "w")
+        else:
+            self.fidx.seek(0)
+
+    def close(self):
+        """Closes the archive and index files."""
+        if not self.fidx.closed:
+            self.fidx.close()
+
+    def reset(self):
+        """Resets the archive and index files."""
+        self.close()
+        self.open()
+
+    @staticmethod
+    def split_name(filepath):
+        """Splits the webdataset into the basename and the extension"""
+        dot_pos = filepath.find(".", filepath.rfind("/") + 1)
+        return filepath[:dot_pos], filepath[dot_pos + 1 :]
+
+    def _get_data_tar(self):
+        """Retreives the data about the offset, name and size of each component
+        using the gnu tar utility, while also filtering out non-file entries"""
+
+        tar_blocks_proc = subprocess.Popen(
+            ["tar", "--list", "--block-num", "--file", self.uri],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        tar_types_sizes_proc = subprocess.Popen(
+            ["tar", "--verbose", "--list", "--file", self.uri],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+
+        tar_blocks = tar_blocks_proc.communicate()[0].split(b"\n")  # block <n>: <filepath>
+        tar_types_sizes = tar_types_sizes_proc.communicate()[0].split(
+            b"\n"
+        )  # <type>... <size> <date> <name>
+
+        # Extracting
+        for blocks_line, types_sizes_line in zip(tar_blocks, tar_types_sizes):
+            if not blocks_line or not types_sizes_line:
+                continue
+
+            name = str(blocks_line[blocks_line.find(b":") + 2 :], "ascii")
+            entry_type = types_sizes_line[0:1]
+
+            if entry_type != b"-":
+                continue
+
+            offset = int(blocks_line[blocks_line.find(b"block") + 6 : blocks_line.find(b":")]) * 512
+            size = types_sizes_line[: -len(name)]
+            size = size[: size.rfind(b"-") - 8]  # "... <size> 20yy-mm-...."
+            size = int(size[size.rfind(b" ") :])
+
+            yield offset, name, size
+
+    def _get_data_tarfile(self):
+        """Retreives the data about the offset, name and size of each component
+        using the tarfile module, while also filtering out non-file entries
+        Intended as a fallback for the gnu tar version (since it is much slower)"""
+
+        print(
+            "Warning: tar utility not found. Falling back to tarfile."
+            + " Processing will most likely take around 100 times longer",
```

Comment:
```suggestion
            + " Processing will most likely take much longer",
```
I don't think we have enough data to claim100 time here.

Reply:
alright